compiler/modules/bir/bir.bal

```diff
@@ -132,18 +132,88 @@ public function lastInsnRef(BasicBlock bb) returns InsnRef {
     return { block: bb.label, index: bb.insns.length() - 1 };
 }
 
-public type Register readonly & record {|
+public enum RegisterKind {
+    PARAM_REGISTER_KIND,
+    VAR_REGISTER_KIND,
+    FINAL_REGISTER_KIND,
+    NARRROW_REGISTER_KIND,
+    TMP_REGISTER_KIND
+}
+
+public type RegisterBase record {|
     # Unique identifier within a function
     # Always >= 0
     int number;
     SemType semType;
-    string? varName;
+    string? name;
     Position? pos;
+    RegisterKind kind;
+|};
+
+public type Register TmpRegister|ParamRegister|VarRegister|FinalRegister|NarrowRegister;
+public type DeclRegisterKind PARAM_REGISTER_KIND|VAR_REGISTER_KIND|FINAL_REGISTER_KIND;
+
+public type DeclRegister record {|
+    *RegisterBase;
+    string name;
+    Position pos;
+    DeclRegisterKind kind;
+|};
+
+public type TmpRegister readonly & record {|
+    *RegisterBase;
+    TMP_REGISTER_KIND kind;
 |};
 
-public function createRegister(FunctionCode code, SemType semType, string? varName = (), Position? pos = ()) returns Register {
+public type NarrowRegister readonly & record {|
+    *RegisterBase;
+    NARRROW_REGISTER_KIND kind;
+|};
+
+public type ParamRegister readonly & record {|
+    *DeclRegister;
+    PARAM_REGISTER_KIND kind;
+|};
+
+public type VarRegister readonly & record {|
+    *DeclRegister;
+    VAR_REGISTER_KIND kind;
+|};
+
+public type FinalRegister readonly & record {|
+    *DeclRegister;
+    FINAL_REGISTER_KIND kind;
+|};
+
+public function createVarRegister(FunctionCode code, SemType semType, string name, Position pos) returns VarRegister {
+    VarRegister r = <VarRegister>createRegister(code, semType, VAR_REGISTER_KIND, name, pos);
```

Comment:
The use cloneReadOnly is ghastly, and the cast is bad. There is no need:

```
public type FinalRegister readonly & record {|
    *DeclRegister;
    FINAL_REGISTER_KIND kind = FINAL_REGISTER_KIND;
|};

public function createFinalRegister(FunctionCode code, SemType semType, string name, Position pos) returns FinalRegister {
    FinalRegister r = { code, semType, name, pos, number: code.registers.length() };
    code.registers.push(r);
    return r;
}
```

Reply:
fixed